### PR TITLE
CIF-2235: Support different url formats for different store configurations

### PR DESCRIPTION
**Pull request by laurentiumagureanu**
## Description

have a custom url format for a store, that is different from the other stores on the same deployment

## Related Issue

CIF-2235

## Motivation and Context

Different SEO needs for different stores

## How Has This Been Tested?

unit test

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes and the overall coverage did not decrease.
- [x] All unit tests pass on CircleCi.
- [x] I ran all tests locally and they pass.

### CODE REVIEW

File: bundles/core/src/main/java/com/adobe/cq/commerce/core/components/services/urls/UrlProvider.java [hunk 1]
```diff
@@ -106,13 +114,21 @@
...
+     * This method should be used if the component only can provide the product
+     * identifier.
+     * 
+     * Either <code>request</code> or <code>request</code> parameter can be
+     * <code>null</code> but not both.
+     * If both are null an {@link IllegalArgumentException} is thrown.
      *
      * @param request The current Sling HTTP request.
-     * @param page The target page, if any. This parameter can be null if the URL template does not use the <code>${page}</code> parameter.
+     * @param page The target page, if any. This parameter can be null
```
Comment (Buuhuu) [inline bundles/core/src/main/java/com/adobe/cq/commerce/core/components/services/urls/UrlProvider.java:129]: Shouldn't that be 

> This parameter can be null if the URL template does set a {{page}} parameter and a request is given?

File: bundles/core/src/main/java/com/adobe/cq/commerce/core/components/services/urls/UrlProvider.java [hunk 2]
```diff
@@ -106,13 +114,21 @@
...
-     * Returns the product page URL. Only the product identifier must be provided, the implementation will query the needed URL
+     * Returns the product page URL. Only the product identifier must be provided,
+     * the implementation will query the needed URL
      * attributes to generate a complete URL based on the configuration.
      * <p>
-     * This method should be used if the component only can provide the product identifier.
+     * This method should be used if the component only can provide the product
+     * identifier.
+     * 
+     * Either <code>request</code> or <code>request</code> parameter can be
```
Comment (Buuhuu) [inline bundles/core/src/main/java/com/adobe/cq/commerce/core/components/services/urls/UrlProvider.java:124]: I guess this means `{@code request} or {@code page}`? Also for all new paragraphs please add a `<p>` tag (no closing one)

File: bundles/core/src/test/java/com/adobe/cq/commerce/core/components/internal/services/UrlProviderImplTest.java [hunk 3]
```diff
@@ -62,11 +71,18 @@
     private MockSlingHttpServletRequest request;
     private CloseableHttpClient httpClient;
     private GraphqlClient graphqlClient;
+    private ComponentsConfiguration caConfig;
+
+    @After
+    public void tearDown() {
```
Comment (Buuhuu) [inline bundles/core/src/test/java/com/adobe/cq/commerce/core/components/internal/services/UrlProviderImplTest.java:77]: Why is this necessary? Each test runs in isolation and gets a new instance of the test class before setup() is called. That means cqConfig will anyway be set to EMPTY.

File: bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/services/UrlProviderImpl.java [hunk 4]
```diff
@@ -94,46 +104,45 @@
...
-    private UrlFormat categoryPageUrlFormat;
+        target = "("
+            + UrlFormat.PROP_USE_AS + "=" + UrlFormat.CATEGORY_PAGE_URL_FORMAT + ")")
+    private List<UrlFormat> categoryPageUrlFormat;
 
-    @Reference(
-        cardinality = ReferenceCardinality.OPTIONAL,
-        policy = ReferencePolicy.STATIC,
-        policyOption = ReferencePolicyOption.GREEDY)
+    @Reference(cardinality = ReferenceCardinality.OPTIONAL, policy = ReferencePolicy.STATIC, policyOption = ReferencePolicyOption.GREEDY)
```
Comment (Buuhuu) [inline bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/services/UrlProviderImpl.java:121]: This should also be `ReferenceCardinality.MULTIPLE`

File: bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/services/UrlProviderImpl.java [hunk 5]
```diff
@@ -94,46 +104,45 @@
...
     private ProductUrlFormat newProductUrlFormat;
 
-    @Reference(
-        cardinality = ReferenceCardinality.OPTIONAL,
-        policy = ReferencePolicy.STATIC,
-        policyOption = ReferencePolicyOption.GREEDY)
+    @Reference(cardinality = ReferenceCardinality.OPTIONAL, policy = ReferencePolicy.STATIC, policyOption = ReferencePolicyOption.GREEDY)
```
Comment (Buuhuu) [inline bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/services/UrlProviderImpl.java:124]: This should also be `ReferenceCardinality.MULTIPLE`

File: bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/services/UrlProviderImpl.java [hunk 6]
```diff
@@ -149,6 +158,69 @@ protected void deactivate() {
         newCategoryUrlFormat = null;
     }
 
+    private ProductUrlFormat getProductFormatUrlFormat(SlingHttpServletRequest request, Page page) {
+        if (request == null && page == null) {
+            throw new IllegalArgumentException("The request and the page parameters cannot both be null");
+        }
+
+        ProductUrlFormat productUrlFormat = newProductUrlFormat;
+        Resource resource = request != null ? request.getResource() : page.getContentResource();
```
Comment (Buuhuu) [inline bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/services/UrlProviderImpl.java:167]: Prefer the `page.getContentResource()` over the `request.getResource()`.

The page is (implicitly) always in the context of the `currentPage` (initially requested page). The resource may be somewhere else due to an include. For XF for example the request.getResource() may point to a different CA Configuration than the page original  request. 

If page != null, use page.getContentResource
else if request != null use request.getResource
else throw IAE

File: bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/services/UrlProviderImpl.java [hunk 7]
```diff
@@ -149,6 +158,69 @@ protected void deactivate() {
...
+            .adaptTo(ComponentsConfiguration.class);
+
+        if (properties != null) {
+            String formatPattern = properties.get(PN_PRODUCT_PAGE_URL_FORMAT, String.class);
+
+            if (formatPattern != null) {
+                if (DEFAULT_PRODUCT_URL_FORMATS.containsKey(formatPattern)) {
+                    productUrlFormat = DEFAULT_PRODUCT_URL_FORMATS.get(formatPattern);
+                } else if (productPageUrlFormat != null && !productPageUrlFormat.isEmpty()) {
+                    UrlFormat customUrlFormat = productPageUrlFormat.stream()
```
Comment (Buuhuu) [inline bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/services/UrlProviderImpl.java:179]: You should also check for the `newProductUrlFormat` containing the custom format.

File: bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/services/UrlProviderImpl.java [hunk 8]
```diff
@@ -149,6 +158,69 @@ protected void deactivate() {
         newCategoryUrlFormat = null;
     }
 
+    private ProductUrlFormat getProductFormatUrlFormat(SlingHttpServletRequest request, Page page) {
```
Comment (Buuhuu) [inline bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/services/UrlProviderImpl.java:161]: This method may be generic to remove duplication
```
private <T> T getFormatFromContext(request, page, propertyName, List<UrlFormat> urlFormats, List<T> newUrlFormats, Function<UrlFormat, T> adapter) {...}
```

File: bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/services/UrlProviderImpl.java [hunk 9]
```diff
@@ -303,22 +369,28 @@ public String getCategoryIdentifier(SlingHttpServletRequest request) {
             return identifier;
         }
 
-        CategoryUrlFormat.Params categoryIdentifiers = newCategoryUrlFormat
-            .parse(request.getRequestPathInfo(), request.getRequestParameterMap());
+        CategoryUrlFormat.Params categoryIdentifiers = getUrlFormatFromContext(request, null,
```
Comment (Buuhuu) [inline bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/services/UrlProviderImpl.java:372]: For better readability, please get the format first and than call `parse()` on it.

File: bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/services/UrlProviderImpl.java [hunk 10]
```diff
@@ -264,23 +325,28 @@ public String getProductIdentifier(SlingHttpServletRequest request) {
...
             MagentoGraphqlClient magentoGraphqlClient = request.adaptTo(MagentoGraphqlClient.class);
             if (magentoGraphqlClient != null) {
                 UrlToProductRetriever productRetriever = new UrlToProductRetriever(magentoGraphqlClient);
                 productRetriever.setIdentifier(productIdentifiers.getUrlKey());
                 ProductInterface product = productRetriever.fetchProduct();
                 identifier = product != null ? product.getSku() : null;
             } else {
-                LOGGER.warn("No backend GraphQL client provided, cannot retrieve product identifier for {}", request.getRequestURL()
-                    .toString());
+                LOGGER.warn("No backend GraphQL client provided, cannot retrieve product identifier for {}",
```
Comment (Buuhuu) [inline bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/services/UrlProviderImpl.java:347]: Please review the formatting changes, there are many new line breaks that are not necessarily needed.

File: bundles/core/src/main/java/com/adobe/cq/commerce/core/components/services/urls/package-info.java [hunk 11]
```diff
@@ -13,7 +13,7 @@
  ~ See the License for the specific language governing permissions and
  ~ limitations under the License.
  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
-@Version("1.2.0")
+@Version("2.0.0")
```
Comment (Buuhuu) [inline bundles/core/src/main/java/com/adobe/cq/commerce/core/components/services/urls/package-info.java:16]: This should not be necessary.

File: bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/services/UrlProviderImpl.java [hunk 12]
```diff
@@ -416,12 +493,22 @@ private String getIdentifierFromRequest(SlingHttpServletRequest request) {
     }
 
     @Override
-    public ProductUrlFormat.Params parseProductUrlFormatParameters(SlingHttpServletRequest request) {
-        return newProductUrlFormat.parse(request.getRequestPathInfo(), request.getRequestParameterMap());
+    public ProductUrlFormat.Params parseProductUrlFormatParameters(SlingHttpServletRequest request, Page page) {
```
Comment (Buuhuu) [inline bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/services/UrlProviderImpl.java:496]: You don't need the `Page` here. Please obtain the context from the request. There are multiple ways you could do that, 

a) get the `currentPage` from the wcm bindings
b) resolve the original requested resource again
c) or just use the request resource

File: bundles/core/src/main/java/com/adobe/cq/commerce/core/components/services/urls/UrlProvider.java [hunk 13]
```diff
@@ -172,9 +219,10 @@
      * configuration for product page URLs.
      *
      * @param request The current Sling HTTP request.
+     * @param page The current page
      * @return the parsed {@link ProductUrlFormat.Params}
      */
-    ProductUrlFormat.Params parseProductUrlFormatParameters(SlingHttpServletRequest request);
+    ProductUrlFormat.Params parseProductUrlFormatParameters(SlingHttpServletRequest request, Page page);
```
Comment (Buuhuu) [inline bundles/core/src/main/java/com/adobe/cq/commerce/core/components/services/urls/UrlProvider.java:225]: Those changes to the API are unnecessary and not required. The context should be fully given by the request.

**Review by Buuhuu — CHANGES_REQUESTED** [review]

File: bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/servlets/AbstractPageUrlFormatsServlet.java [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+import org.apache.sling.api.resource.ValueMap;
+import org.apache.sling.api.servlets.SlingSafeMethodsServlet;
+import org.apache.sling.api.wrappers.ValueMapDecorator;
+
+import com.adobe.cq.commerce.core.components.services.urls.UrlFormat;
+import com.adobe.granite.ui.components.ds.DataSource;
+import com.adobe.granite.ui.components.ds.SimpleDataSource;
+import com.adobe.granite.ui.components.ds.ValueMapResource;
+
+public class AbstractPageUrlFormatsServlet extends SlingSafeMethodsServlet {
```
Comment (Buuhuu) [inline bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/servlets/AbstractPageUrlFormatsServlet.java:35]: make the class abstract if it is called "Abstract...Servlet"

File: bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/servlets/CategoryPageUrlFormatsServlet.java [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+        CategoryPageWithUrlKey.PATTERN.replace("#", "\\#"),
+        CategoryPageWithUrlPath.PATTERN.replace("#", "\\#"));
+
+    @Reference(
+        cardinality = ReferenceCardinality.MULTIPLE,
+        policy = ReferencePolicy.STATIC,
+        policyOption = ReferencePolicyOption.GREEDY,
+        target = "("
+            + UrlFormat.PROP_USE_AS + "=" + UrlFormat.CATEGORY_PAGE_URL_FORMAT + ")")
+    private List<UrlFormat> categoryPageUrlFormats;
```
Comment (Buuhuu) [inline bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/servlets/CategoryPageUrlFormatsServlet.java:56]: You should also include the new, typed url formats here. `UrlFromat` (untyped) itself is deprecated.

File: bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/servlets/ProductPageUrlFormatsServlet.java [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+        ProductPageWithUrlKey.PATTERN.replace("#", "\\#"),
+        ProductPageWithUrlPath.PATTERN.replace("#", "\\#"));
+
+    @Reference(
+        cardinality = ReferenceCardinality.MULTIPLE,
+        policy = ReferencePolicy.STATIC,
+        policyOption = ReferencePolicyOption.GREEDY,
+        target = "("
+            + UrlFormat.PROP_USE_AS + "=" + UrlFormat.PRODUCT_PAGE_URL_FORMAT + ")")
+    private List<UrlFormat> productPageUrlFormats;
```
Comment (Buuhuu) [inline bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/servlets/ProductPageUrlFormatsServlet.java:62]: You should also include the new, typed url formats here. UrlFromat (untyped) itself is deprecated.

File: bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/servlets/ProductPageUrlFormatsServlet.java [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+    service = { Servlet.class },
+    property = {
+        "sling.servlet.resourceTypes=" + ProductPageUrlFormatsServlet.RESOURCE_TYPE,
+        "sling.servlet.methods=GET",
+        "sling.servlet.extensions=html"
+    })
+public class ProductPageUrlFormatsServlet extends AbstractPageUrlFormatsServlet {
+    static final String RESOURCE_TYPE = "core/cif/components/datasource/producturlformats";
+    static final String PAGE_URL_FORMAT_KEY = "productPageUrlFormat";
+    static final List<String> URL_FORMATS = Arrays.asList(
```
Comment (Buuhuu) [inline bundles/core/src/main/java/com/adobe/cq/commerce/core/components/internal/servlets/ProductPageUrlFormatsServlet.java:49]: It would be better to add a method `getAllProductPageUrlFromats()` to the `UrlProviderImpl` (only the implementation!) and add a reference from the servlet to the `UrlProviderImpl`. So you would not have to duplicate the "logic" of binding to the custom formats and you could use the statically defined default formats.

**Review by Buuhuu — CHANGES_REQUESTED** [review]

File: bundles/core/src/main/java/com/adobe/cq/commerce/core/components/services/urls/package-info.java (hunk 11 above)
Comment (laurentiumagureanu) [reply bundles/core/src/main/java/com/adobe/cq/commerce/core/components/services/urls/package-info.java:16]: does not build without it. because I changed a method signature

**Review by Buuhuu — APPROVED** [review]

**Review by herzog31 — APPROVED** [review]
